### DISCUSSION TITLE
V3 - Industrial Communication Activities

### DISCUSSION THREAD

**Discussion by KnibbsyMan** [elsa-workflows/elsa-core#4625] [Q&A]
I work in the manufacturing industry and have been looking at using Elsa Workflows to help simplify some of the repetitive tasks we have to do and to automate certain activities. My question is, would including support for industrial ethernet communications protocols, i.e. OPC UA or Modbus, be within the scope of what you are doing with the project? I can see there is already a growing selection of activities types, so I was wondered if this is something you would consider adding to the project? Being able to ingest data directly from shop floor events or tag value changes would be really useful in some of my use cases. 

An example of this could be, when the temperature of a  sensors value goes above X, log it in a database and send an email to management to warn them of the issue.

**Comment by sfmskywalker** [comment]
Absolutely, your suggestion for integrating such protocols into Elsa is both insightful and relevant, especially for applications in the manufacturing sector. I envision Elsa as a platform that supports a wide range of activities, including those specific to industrial automation.

While these specialized protocols may be outside the immediate expertise of the core Elsa team, we highly encourage and support community contributions to extend our functionalities. Modules like the ones you propose could ideally be developed and maintained in a community repository. This collaborative approach allows experts like yourself to contribute specific knowledge, enriching the Elsa ecosystem.

In the future, we might establish a marketplace website for easier discovery of these modules. This could include both free and premium modules, with premium modules potentially residing in vendor-specific repositories, like Elsa Digital (the commercial branch around Elsa), which will contain curated, high-quality modules. Community-contributed modules will be housed in a separate "community" repository.

We're keen on fostering a collaborative environment where users from various industries can contribute. Your use case of triggering actions based on sensor data is a great example of the potential applications of Elsa in industrial settings.

**Reply by KnibbsyMan** [reply]
Thanks for taking the time to get back to me. I am really glad to hear this. I have a real interest in trying to integrate Elsa 3 into our facility with its existing capabilities, to try and resolve some of our repetitive tasks, as previously mentioned. 

When V3 is officially released (in the next couple of weeks by the looks of it?) I will try and allocate some time to understand how Elsa works behind the scenes and how the existing activities / modules are designed. I’ll then try to create an Elsa library for OPC UA, Modbus or other, based on other examples. If I manage to make any headway with one I will try and submit it here, to the community, if you’re happy with that? I should also add, I’m no expert in all things C# but I am happy to try and contribute my development towards this project, if they would be useful.

On the development note, I assume using other open source libraries within your my solution would be allowed in my implementation? As packages like NModbus and OPCFoundation.NetStandard.Opc.Ua are already available for these protocols and would do a lot of the heavy lifting for me. It might just then be a case of leveraging them and wrapping them in Elsa activities, which would drastically simplify things for me.

My only concern / question I have with the OPC Foundation is that they are licence under GPLv2, which if I’m not mistaken, forces all projects using this package to also be under said licence. If so, would this mean we couldn’t add support for this protocol or have I misunderstood what I’ve read about this licence?
